File: packages/contracts/rest.py
"""REST API Request and Response Contracts."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from .enums import (
    DegradationTier,
    MeetingStatus,
    ParticipantRole,
    TranscriptFormat,
    WorkerHealthStatus,
)


class BaseContract(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,
        validate_assignment=True,
        extra="forbid",
    )


class AuthTokenRequest(BaseContract):
    user_id: str = Field(..., description="Unique user identifier")
    tenant_id: str = Field(..., description="Organization or tenant identifier")
    email: str = Field(..., description="User primary email address")
    role: ParticipantRole = Field(default=ParticipantRole.PARTICIPANT)


class AuthTokenResponse(BaseContract):
    access_token: str
    token_type: str = "Bearer"
    expires_in_sec: int
    tenant_id: str
    user_id: str


class CreateMeetingRequest(BaseContract):
    title: str = Field(..., min_length=1, max_length=200, description="Meeting title")
    host_spoken_language: str = Field(
        default="eng", min_length=3, max_length=3, description="ISO-639-3 spoken language"
    )
    host_listening_language: str = Field(
        default="eng", min_length=3, max_length=3, description="ISO-639-3 listening language"
    )
    scheduled_start: datetime | None = None
    passcode: str | None = Field(default=None, max_length=32)


class CreateMeetingResponse(BaseContract):
    meeting_id: str
    tenant_id: str
    title: str
    status: MeetingStatus = MeetingStatus.SCHEDULED
    state_version: int = 1
    created_at: datetime


class GetMeetingResponse(BaseContract):
    meeting_id: str
    tenant_id: str
    title: str
    status: MeetingStatus
    state_version: int
    created_at: datetime
    active_participants_count: int = 0


class EndMeetingRequest(BaseContract):
    reason: str = Field(default="HOST_TERMINATED", max_length=200)


class JoinMeetingRequest(BaseContract):
    display_name: str = Field(..., min_length=1, max_length=100)
    spoken_language: str = Field(
        default="eng", min_length=3, max_length=3, description="ISO-639-3 code"
    )
    listening_language: str = Field(
        default="eng", min_length=3, max_length=3, description="ISO-639-3 code"
    )
    passcode: str | None = None


class JoinMeetingResponse(BaseContract):
    meeting_id: str
    participant_id: str
    display_name: str
    role: ParticipantRole
    livekit_token: str
    ws_ticket: str
    state_version: int


class UpdateParticipantRequest(BaseContract):
    display_name: str | None = Field(default=None, max_length=100)
    spoken_language: str | None = Field(default=None, min_length=3, max_length=3)
    listening_language: str | None = Field(default=None, min_length=3, max_length=3)
    is_muted: bool | None = None
    is_video_enabled: bool | None = None


class ParticipantContract(BaseContract):
    participant_id: str
    user_id: str | None = None
    display_name: str
    role: ParticipantRole
    spoken_language: str = Field(..., description="ISO-639-3 spoken language code")
    listening_language: str = Field(..., description="ISO-639-3 personalized listening language")
    is_muted: bool = False
    is_video_enabled: bool = True
    joined_at: datetime


class MeetingContract(BaseContract):
    meeting_id: str
    tenant_id: str
    title: str
    status: MeetingStatus
    state_version: int = 1
    created_at: datetime
    updated_at: datetime


class TranscriptSegmentResponse(BaseContract):
    source_segment_id: str
    speaker_id: str
    speaker_name: str
    source_language: str
    target_language: str
    original_text: str
    translated_text: str
    start_ms: int
    end_ms: int
    is_final: bool


class GetTranscriptResponse(BaseContract):
    meeting_id: str
    total_segments: int
    format: TranscriptFormat
    segments: list[TranscriptSegmentResponse]


class OrganizationResponse(BaseContract):
    id: str
    name: str
    slug: str
    created_at: datetime
    member_count: int = 0
    meeting_count: int = 0


class UpdateOrganizationRequest(BaseContract):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    slug: str | None = Field(default=None, min_length=1, max_length=100)


class OrganizationMemberResponse(BaseContract):
    user_id: str
    email: str
    role: ParticipantRole
    display_name: str | None = None
    is_active: bool = True
    created_at: datetime


class InviteMemberRequest(BaseContract):
    email: str = Field(..., min_length=3, max_length=255)
    role: ParticipantRole = Field(default=ParticipantRole.PARTICIPANT)
    display_name: str | None = Field(default=None, max_length=100)


class UpdateMemberRoleRequest(BaseContract):
    role: ParticipantRole | None = None
    is_active: bool | None = None


class AdminMeetingSummaryResponse(BaseContract):
    meeting_id: str
    title: str
    status: MeetingStatus
    scheduled_start: datetime | None = None
    started_at: datetime | None = None
    ended_at: datetime | None = None
    duration_seconds: int | None = None
    participant_count: int = 0
    transcript_segment_count: int = 0
    created_at: datetime


class AdminAnalyticsResponse(BaseContract):
    tenant_id: str
    active_meetings_count: int
    total_meetings_count: int
    total_transcribed_minutes: float
    total_participants_count: int
    language_breakdown: dict[str, int]
    average_translation_latency_ms: float


class AuditLogEntry(BaseContract):
    id: str
    event_type: str
    actor_email: str
    target: str
    timestamp: datetime
    details: dict[str, Any] = Field(default_factory=dict)


class AdminAuditLogsResponse(BaseContract):
    logs: list[AuditLogEntry]
    total: int


class WorkerHeartbeatPayload(BaseContract):
    worker_type: str = Field(..., description="Worker service type (stt, nmt, tts, speaker, assistant)")
    worker_id: str = Field(..., description="Unique instance identifier of the worker")
    timestamp_ms: int = Field(..., description="Epoch millisecond timestamp of the heartbeat")
    queue_depth: int = Field(default=0, ge=0, description="Observed queue depth for this worker")
    gpu_utilization_pct: float | None = Field(default=None, ge=0.0, le=100.0, description="Optional GPU load %")


class PipelineStatusResponse(BaseContract):
    meeting_id: str = Field(..., description="Unique meeting room identifier")
    current_tier: DegradationTier = Field(default=DegradationTier.NORMAL)
    active_workers: dict[str, WorkerHealthStatus] = Field(default_factory=dict)
    queue_depths: dict[str, int] = Field(default_factory=dict)
    dropped_partials_count: int = Field(default=0, ge=0)
    uptime_seconds: float = Field(default=0.0, ge=0.0)

File: packages/contracts/src/rest.ts
import {
  DegradationTier,
  MeetingStatus,
  ParticipantRole,
  TranscriptFormat,
  WorkerHealthStatus,
} from "./enums";

export interface AuthTokenRequest {
  user_id: string;
  tenant_id: string;
  email: string;
  role?: ParticipantRole;
}

export interface AuthTokenResponse {
  access_token: string;
  token_type: string;
  expires_in_sec: number;
  tenant_id: string;
  user_id: string;
}

export interface CreateMeetingRequest {
  title: string;
  host_spoken_language?: string;
  host_listening_language?: string;
  scheduled_start?: string;
  passcode?: string;
}

export interface CreateMeetingResponse {
  meeting_id: string;
  tenant_id: string;
  title: string;
  status: MeetingStatus;
  state_version: number;
  created_at: string;
}

export interface GetMeetingResponse {
  meeting_id: string;
  tenant_id: string;
  title: string;
  status: MeetingStatus;
  state_version: number;
  created_at: string;
  active_participants_count: number;
}

export interface EndMeetingRequest {
  reason?: string;
}

export interface JoinMeetingRequest {
  display_name: string;
  spoken_language?: string;
  listening_language?: string;
  passcode?: string;
}

export interface JoinMeetingResponse {
  meeting_id: string;
  participant_id: string;
  display_name: string;
  role: ParticipantRole;
  livekit_token: string;
  ws_ticket: string;
  state_version: number;
}

export interface UpdateParticipantRequest {
  display_name?: string;
  spoken_language?: string;
  listening_language?: string;
  is_muted?: boolean;
  is_video_enabled?: boolean;
}

export interface ParticipantContract {
  participant_id: string;
  user_id?: string | null;
  display_name: string;
  role: ParticipantRole;
  spoken_language: string;
  listening_language: string;
  is_muted: boolean;
  is_video_enabled: boolean;
  joined_at: string;
}

export interface MeetingContract {
  meeting_id: string;
  tenant_id: string;
  title: string;
  status: MeetingStatus;
  state_version: number;
  created_at: string;
  updated_at: string;
}

export interface TranscriptSegmentResponse {
  source_segment_id: string;
  speaker_id: string;
  speaker_name: string;
  source_language: string;
  target_language: string;
  original_text: string;
  translated_text: string;
  start_ms: number;
  end_ms: number;
  is_final: boolean;
}

export interface GetTranscriptResponse {
  meeting_id: string;
  total_segments: number;
  format: TranscriptFormat;
  segments: TranscriptSegmentResponse[];
}

export interface OrganizationResponse {
  id: string;
  name: string;
  slug: string;
  created_at: string;
  member_count: number;
  meeting_count: number;
}

export interface UpdateOrganizationRequest {
  name?: string;
  slug?: string;
}

export interface OrganizationMemberResponse {
  user_id: string;
  email: string;
  role: ParticipantRole;
  display_name?: string | null;
  is_active: boolean;
  created_at: string;
}

export interface InviteMemberRequest {
  email: string;
  role?: ParticipantRole;
  display_name?: string | null;
}

export interface UpdateMemberRoleRequest {
  role?: ParticipantRole;
  is_active?: boolean;
}

export interface AdminMeetingSummaryResponse {
  meeting_id: string;
  title: string;
  status: MeetingStatus;
  scheduled_start?: string | null;
  started_at?: string | null;
  ended_at?: string | null;
  duration_seconds?: number | null;
  participant_count: number;
  transcript_segment_count: number;
  created_at: string;
}

export interface AdminAnalyticsResponse {
  tenant_id: string;
  active_meetings_count: number;
  total_meetings_count: number;
  total_transcribed_minutes: number;
  total_participants_count: number;
  language_breakdown: Record<string, number>;
  average_translation_latency_ms: number;
}

export interface AuditLogEntry {
  id: string;
  event_type: string;
  actor_email: string;
  target: string;
  timestamp: string;
  details: Record<string, unknown>;
}

export interface AdminAuditLogsResponse {
  logs: AuditLogEntry[];
  total: number;
}

export interface WorkerHeartbeatPayload {
  worker_type: string;
  worker_id: string;
  timestamp_ms: number;
  queue_depth?: number;
  gpu_utilization_pct?: number | null;
}

export interface PipelineStatusResponse {
  meeting_id: string;
  current_tier: DegradationTier;
  active_workers: Record<string, WorkerHealthStatus>;
  queue_depths: Record<string, number>;
  dropped_partials_count: number;
  uptime_seconds: number;
}
